File: examples/python/function-patterns/FunctionPatterns.runar.py
from runar import (
    StatefulSmartContract, PubKey, Sig, Bigint, Readonly,
    public, assert_, check_sig, percent_of, mul_div, clamp, safemod,
)

class FunctionPatterns(StatefulSmartContract):
    owner: Readonly[PubKey]
    balance: Bigint

    def __init__(self, owner: PubKey, balance: Bigint):
        super().__init__(owner, balance)
        self.owner = owner
        self.balance = balance

    @public
    def deposit(self, sig: Sig, amount: Bigint):
        self._require_owner(sig)
        assert_(amount > 0)
        self.balance = self.balance + amount

    @public
    def withdraw(self, sig: Sig, amount: Bigint, fee_bps: Bigint):
        self._require_owner(sig)
        assert_(amount > 0)
        fee = self._compute_fee(amount, fee_bps)
        total = amount + fee
        assert_(total <= self.balance)
        self.balance = self.balance - total

    @public
    def scale(self, sig: Sig, numerator: Bigint, denominator: Bigint):
        self._require_owner(sig)
        self.balance = self._scale_value(self.balance, numerator, denominator)

    @public
    def normalize(self, sig: Sig, lo: Bigint, hi: Bigint, step: Bigint):
        self._require_owner(sig)
        clamped = self._clamp_value(self.balance, lo, hi)
        self.balance = self._round_down(clamped, step)

    def _require_owner(self, sig: Sig):
        assert_(check_sig(sig, self.owner))

    def _compute_fee(self, amount: Bigint, fee_bps: Bigint) -> Bigint:
        return percent_of(amount, fee_bps)

    def _scale_value(self, value: Bigint, numerator: Bigint, denominator: Bigint) -> Bigint:
        return mul_div(value, numerator, denominator)

    def _clamp_value(self, value: Bigint, lo: Bigint, hi: Bigint) -> Bigint:
        return clamp(value, lo, hi)

    def _round_down(self, value: Bigint, step: Bigint) -> Bigint:
        remainder = safemod(value, step)
        return value - remainder

File: examples/ruby/function-patterns/FunctionPatterns.runar.rb
require 'runar'

class FunctionPatterns < Runar::StatefulSmartContract
  prop :owner, PubKey, readonly: true
  prop :balance, Bigint

  def initialize(owner, balance)
    super(owner, balance)
    @owner = owner
    @balance = balance
  end

  runar_public sig: Sig, amount: Bigint
  def deposit(sig, amount)
    _require_owner(sig)
    assert amount > 0
    @balance = @balance + amount
  end

  runar_public sig: Sig, amount: Bigint, fee_bps: Bigint
  def withdraw(sig, amount, fee_bps)
    _require_owner(sig)
    assert amount > 0
    fee = _compute_fee(amount, fee_bps)
    total = amount + fee
    assert total <= @balance
    @balance = @balance - total
  end

  runar_public sig: Sig, numerator: Bigint, denominator: Bigint
  def scale(sig, numerator, denominator)
    _require_owner(sig)
    @balance = _scale_value(@balance, numerator, denominator)
  end

  runar_public sig: Sig, lo: Bigint, hi: Bigint, step: Bigint
  def normalize(sig, lo, hi, step)
    _require_owner(sig)
    clamped = _clamp_value(@balance, lo, hi)
    @balance = _round_down(clamped, step)
  end

  private

  params sig: Sig
  def _require_owner(sig)
    assert check_sig(sig, @owner)
  end

  params amount: Bigint, fee_bps: Bigint
  def _compute_fee(amount, fee_bps)
    percent_of(amount, fee_bps)
  end

  params value: Bigint, numerator: Bigint, denominator: Bigint
  def _scale_value(value, numerator, denominator)
    mul_div(value, numerator, denominator)
  end

  params value: Bigint, lo: Bigint, hi: Bigint
  def _clamp_value(value, lo, hi)
    clamp(value, lo, hi)
  end

  params value: Bigint, step: Bigint
  def _round_down(value, step)
    remainder = safemod(value, step)
    value - remainder
  end
end
